### PR TITLE
fix(email): skip MJML theme HTML fallback when customHtml empty

### LINKED ISSUE #16988: Emails created programmatically with an MJML theme and empty customHtml are sent as raw, uncompiled MJML in the text/html part
## Description

When an email entity has an empty `customHtml` and a `template` set, `MailHelper::setEmail()` falls back to rendering the theme's `email.html.twig` and uses the result directly as the `text/html` body.

Since Mautic 5 the bundled themes are **MJML**, not HTML. `themes/blank/html/email.html.twig` starts with `<mjml>`. There is no server-side MJML compilation in this path — the MJML→HTML compilation only happens client-side, in the GrapesJS builder, which stores the compiled HTML into `custom_html`.

The result is that the recipient receives raw, uncompiled MJML as the `text/html` MIME part:

```
Content-Type: text/html; charset=utf-8

<mjml>
    <mj-head>
      <mj-attributes>
      </mj-attributes>
      ...
      <mj-text font-size="28px" font-weight="700">
        <p>Hello World!</p>
      </mj-text>
```

Mail clients ignore the unknown `mj-*` tags, so the recipient sees the bare text of the boilerplate ("Hello World! / Lorem ipsum…") with no layout, instead of the intended content.

## Code path

`app/bundles/EmailBundle/Helper/MailHelper.php` (7.x, currently `7.2.0-rc2`):

```php
$template   = $email->getTemplate();
$customHtml = $email->getCustomHtml();
// Process emails created by Mautic v1
if (empty($customHtml) && $template) {
    $logicalName = $this->themeHelper->checkForTwigTemplate('@themes/'.$template.'/html/email.html.twig');

    $customHtml = $this->themeHelper->renderThemeTemplate($logicalName, [
        'content'  => $email->getContent(),
        'email'    => $email,
        'template' => $template,
    ]);
}

$this->setBody($customHtml, 'text/html', null, $ignoreTrackingPixel);
```

The comment (`Process emails created by Mautic v1`) suggests this fallback predates MJML themes, when rendering a theme template produced valid HTML. With the MJML themes shipped today it produces an invalid HTML body, and nothing guards against it.

## Steps to reproduce

1. Create an email through the API (or any programmatic path) with a theme but without HTML content:

```http
POST /api/emails/new
{
  "name": "MJML fallback test",
  "subject": "MJML fallback test",
  "emailType": "template",
  "template": "blank",
  "customHtml": "",
  "plainText": "Plain text body."
}
```

2. Send it to a real address.
3. Inspect the raw source of the received message.

## Expected

The `text/html` part contains valid HTML — either by compiling the theme MJML server-side, or by not producing an HTML part at all when no valid HTML can be produced.

## Actual

The `text/html` part contains the raw MJML source of `themes/blank/html/email.html.twig`, with theme tokens (`{unsubscribe_text}`, `{webview_text}`) replaced but the MJML never compiled.

Note that this happens **after** the entity is saved: the stored `custom_html` stays empty, and the MJML is injected at send time by the fallback above. So the email looks empty/fine in the database and only breaks on the wire.

## Environment

- Mautic 7.1.3 (reproduced); fallback code unchanged on `7.x` at `7.2.0-rc2`
- Theme: `blank` (bundled)
- Any transport — the fallback runs before the transport is involved

## Workarounds

- Provide a compiled `customHtml` when creating the email programmatically
- Create/save the email through the UI once, so GrapesJS compiles the MJML into `custom_html`
- Use a non-MJML theme

## Possible fixes

Some options, in increasing order of effort:

1. Detect `<mjml>` in the rendered template and skip the fallback (leave the HTML part out) rather than shipping invalid HTML.
2. Log an error / surface a validation error when the fallback yields MJML, so the failure is not silent.
3. Compile the MJML server-side in this path. The builder already has a server-side compilation endpoint used by `MjmlService.mjmlToHtmlViaEndpoint()`, so the capability exists in the codebase.

I am happy to work on a PR if maintainers agree on which direction is preferred.

### PR DESCRIPTION
**Pull request by santhiprakash**
| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | ✔️
| New feature/enhancement? (use the a.x branch)      | ❌
| Deprecations?                          | ❌
| BC breaks? (use the c.x branch)        | ❌
| Automated tests included?              | ✔️
| Related user documentation PR URL      | n/a
| Related developer documentation PR URL | n/a
| Issue(s) addressed                     | Fixes #16988

## Problem

When an email is created programmatically (REST API) with a theme/`template` set but empty `customHtml`, `MailHelper::setEmail()` falls back to rendering the theme's `html/email.html.twig` and sends that result directly as the `text/html` MIME part.

Since Mautic 5, bundled themes are MJML (`themes/blank/html/email.html.twig` starts with `<mjml>`). MJML→HTML compilation only happens client-side in GrapesJS and is stored in `custom_html`. On this send-time fallback path there is no compilation, so recipients receive raw uncompiled MJML as `text/html` and see bare boilerplate text.

## Triage / Root cause

In `MailHelper::setEmail()`, the Mautic v1 theme fallback runs whenever `customHtml` is empty and a template is set, then always calls `setBody($customHtml, 'text/html')` with the rendered theme output — including when that output is MJML.

## Fix

- Detect `<mjml>` in the rendered theme fallback (same signal GrapesJS already uses).
- Skip assigning that render as the HTML body.
- Log a warning so the silent failure is visible.
- Preserve the legacy HTML-theme fallback path for non-MJML themes.

When HTML is skipped and `plainText` is present, existing send logic already omits the empty HTML part (`!empty($body['content']) || empty($plainText)`).

## Verification

```bash
docker run --rm -v "$PWD":/app -w /app php:8.3-cli \
  php -d zend.assertions=1 bin/phpunit -d memory_limit=512M \
  --bootstrap vendor/autoload.php --configuration app/phpunit.xml.dist \
  --filter 'testSetEmailSkipsMjmlThemeFallbackWhenCustomHtmlEmpty|testSetEmailStillUsesHtmlThemeFallbackWhenCustomHtmlEmpty' \
  app/bundles/EmailBundle/Tests/Helper/MailHelperTest.php
```

Result: OK (2 tests, 7 assertions).

Also: `php-cs-fixer` dry-run clean on the touched files.

## Notes / Risks

- This is the smallest correct guard from the issue options (skip invalid HTML rather than adding server-side MJML compilation).
- Emails that only had MJML theme boilerplate and no `customHtml`/`plainText` will now send without an HTML body; API/UI creators should supply compiled `customHtml` (or open the email in the builder once).
- Targeted `7.x` to match current merge practice for this class of bug and the issue's 7.x repro; happy to retarget to a lower maintained branch if maintainers prefer.

---
### 📋 Steps to test this PR:

1. Open this PR on GitHub Codespaces or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. Create an email via API with an MJML theme and empty HTML:

```http
POST /api/emails/new
{
  "name": "MJML fallback test",
  "subject": "MJML fallback test",
  "emailType": "template",
  "template": "blank",
  "customHtml": "",
  "plainText": "Plain text body."
}
```

3. Send it to a real address (or inspect the generated MIME before transport).
4. Confirm the `text/html` part is absent/empty and does **not** contain raw `<mjml>` markup; plain text still sends.
5. Optionally create a similar email with a non-MJML/legacy HTML theme and empty `customHtml` and confirm that theme HTML fallback still works.

### CODE REVIEW

**Comment by santhiprakash** [comment]
## CI follow-up

Pushed `fd7d32c25c` to address the failing checks from the first run:

1. **Rector** — removed redundant `(string)` casts (`RecastingRemovalRector`) and decorated `willReturnMap` with `expects($this->atLeast(4))` (`DecorateWillReturnMapWithExpectsMockRector`).
2. **PHPUnit** (`EmailControllerFunctionalTest::testSendEmailForImportCustomEmailTemplate`) — blank theme is MJML; with empty `customHtml`, the new guard correctly leaves the import body empty instead of returning raw `<mjml>`. Updated the assertion and normalized skipped fallback `customHtml` to `''`.

Local verification: `docker run php:8.3-cli bin/phpunit --filter testSetEmailSkipsMjmlThemeFallbackWhenCustomHtmlEmpty|testSetEmailStillUsesHtmlThemeFallbackWhenCustomHtmlEmpty` → OK (2 tests). Rector dry-run on the touched MailHelper files → clean.

File: app/bundles/EmailBundle/Helper/MailHelper.php [hunk 1]
```diff
@@ -1257,11 +1257,25 @@ public function setEmail(Email $email, $allowBcc = true, $assetAttachments = [],
...
+            // Bundled themes since Mautic 5 are MJML. GrapesJS compiles them into
+            // customHtml client-side; sending the raw theme template as text/html
+            // delivers uncompiled <mjml> markup to recipients.
+            if (str_contains($renderedHtml, '<mjml>')) {
+                $this->logger->warning(sprintf(
+                    'Email%s uses MJML theme "%s" with empty customHtml; skipping theme HTML fallback to avoid sending uncompiled MJML.',
+                    null !== $email->getId() ? ' ID '.$email->getId() : '',
+                    $template
+                ));
+                $customHtml = '';
```
Comment (peculiaruc) [inline app/bundles/EmailBundle/Helper/MailHelper.php:1275]: If MJML fallback is skipped and there is no `plainText` either, please make `setEmail()` return `false` via `$this->errors` rather than only warning. Otherwise send still proceeds with an empty body.

File: app/bundles/EmailBundle/Tests/Helper/MailHelperTest.php [hunk 2]
```diff
@@ -1390,6 +1390,72 @@ public function testSetBccWithAssociativeArray(): void
         $this->assertEquals('Default Name', $bcc[1]->getName());
     }
 
+    public function testSetEmailSkipsMjmlThemeFallbackWhenCustomHtmlEmpty(): void
+    {
```
Comment (peculiaruc) [inline app/bundles/EmailBundle/Tests/Helper/MailHelperTest.php:1394]: Can we also:
1. assert `$this->logger` receives the warning once
2. add a case with empty `customHtml` + empty `plainText` (expected: hard fail once that behavior is added)

File: app/bundles/EmailBundle/Tests/Controller/EmailControllerFunctionalTest.php [hunk 3]
```diff
@@ -1211,7 +1211,9 @@ public function testSendEmailForImportCustomEmailTemplate(): void
         $this->assertSame(1, $response['success']);
         $this->assertNotEmpty($response['subject']);
         $this->assertEquals($email->getSubject(), $response['subject']);
-        $this->assertNotEmpty($response['body']);
+        // Blank theme is MJML; with empty customHtml, MailHelper skips the theme
+        // fallback so the import body stays empty rather than raw uncompiled <mjml>.
+        $this->assertSame('', $response['body']);
```
Comment (peculiaruc) [inline app/bundles/EmailBundle/Tests/Controller/EmailControllerFunctionalTest.php:1216]: You can add an API/functional test that creates an email with an MJML theme and empty `customHtml`, then sends or previews it and asserts the body/MIME never contains `<mjml>`. That would cover the #16988 regression end-to-end, not only `MailHelper::setEmail()`.

**Review by peculiaruc — CHANGES_REQUESTED** [review]
Thank you for your PR.

You correctly implement option 1 from #16988 and stop raw `<mjml>` being sent as `text/html` when `customHtml` is empty.

But your PR does not fully close the issue yet: empty MJML-theme emails can still be created and successfully sent with no useful content (only a log warning).

Could you please address the 1-3 before we merge?
1. Hard-fail empty content after MJML skip when both HTML and plain text are empty (don’t only log + continue).
2. Create/save validation for MJML theme + empty `customHtml` (API/UI), so the bad state is blocked earlier.
3. And the inline suggestions.

**Comment by santhiprakash** [comment]
Thanks for the thorough review @peculiaruc — all three asks are addressed in `c60624fb3e`:

**1. Hard-fail empty content after MJML skip**
`MailHelper::setEmail()` now populates `$this->errors` when the MJML theme fallback is skipped and `plainText` is also empty, so `setEmail()` returns `false` and the send is aborted rather than delivering an empty body.

**2. Create/save validation for MJML theme + empty `customHtml`**
Added `EmailModel::validateMjmlThemeHasCustomHtml()` called from `saveEntity()`. It renders the theme template and checks for `<mjml>`; if found with empty `customHtml`, it throws a new `MjmlThemeEmptyCustomHtmlException`.
- **API**: `EmailApiController::preSaveEntity()` catches it and returns a 422 response.
- **UI**: `EmailController` catches it in new/edit/clone actions and shows a flash error.

**3. Inline suggestions**
- `MailHelper.php` ~L1275: `setEmail()` now returns `false` via `$this->errors` when both HTML and plain text are empty. ✓
- `MailHelperTest.php` ~L1394: the existing test now asserts `$this->logger` receives the warning exactly once; added `testSetEmailFailsWhenMjmlThemeSkippedAndNoPlainText` covering empty `customHtml` + empty `plainText` (expects `assertFalse` + non-empty errors). ✓
- `EmailControllerFunctionalTest.php` ~L1216: added `testSaveRejectsMjmlThemeWithEmptyCustomHtml` — creates an email via `POST /api/emails/new` with `template: blank` + empty `customHtml` and asserts a 422 response; the existing import test also now asserts the body never contains `<mjml>`. ✓

All targeted tests pass:
```
MailHelperTest: 3 tests, 18 assertions — OK
EmailControllerFunctionalTest: 2 tests, 11 assertions — OK
```

**Comment by santhiprakash** [comment]
## CI fix follow-up

The previous push (`c60624fb3e`) failed PHPUnit because `validateMjmlThemeHasCustomHtml()` was placed inside `EmailModel::saveEntity()`, which is called by every save path — including test fixtures, campaign imports, and project associations. This broke 17 unrelated tests that legitimately save MJML-themed emails with empty `customHtml` as fixtures or drafts.

Fixed in `3b6a347f29`:

- **Removed** the validation call from `EmailModel::saveEntity()`. Programmatic saves are no longer blocked; the send-time hard-fail in `MailHelper::setEmail()` remains as the final safety net.
- **Scoped** `validateMjmlThemeHasCustomHtml()` to published emails only (`!$email->getIsPublished()` early return). Unpublished/draft emails can be saved with an MJML theme and empty `customHtml` — the user opens the builder to compile before publishing.
- **Moved** the validation call into `EmailController` (new/edit/clone) before `saveEntity()`, inside the existing try blocks. The API controller's `preSaveEntity` already calls it. This way only user-facing saves (UI/API) are validated, not programmatic ones.
- **Updated** `testPublishPermissionOnNewEmailForAdminUser` and `testPublishPermissionOnCreate` to set `customHtml`, since they create published emails with the blank (MJML) template through the UI, which is now correctly blocked.

All 15 CI checks now pass.

File: app/bundles/EmailBundle/Model/EmailModel.php [hunk 4]
```diff
@@ -274,6 +275,44 @@ public function saveEntity($entity, $unlock = true): void
...
+        $logicalName = $this->themeHelper->checkForTwigTemplate('@themes/'.$template.'/html/email.html.twig');
+
+        try {
+            $renderedHtml = $this->themeHelper->renderThemeTemplate($logicalName, [
+                'content'  => $email->getContent(),
+                'email'    => $email,
+                'template' => $template,
+            ]);
+        } catch (\Throwable) {
+            // If the theme template cannot be rendered, skip validation —
```
Comment (andersonjeccel) [inline app/bundles/EmailBundle/Model/EmailModel.php:314]: Could we catch the specific template-rendering exception here instead? Swallowing every Throwable lets a broken theme save with empty customHtml, so the send-time fallback can still fail later.

File: app/bundles/EmailBundle/Model/EmailModel.php [hunk 5]
```diff
@@ -274,6 +275,44 @@ public function saveEntity($entity, $unlock = true): void
...
+     * customHtml client-side. If a published email is saved with an MJML theme
+     * but empty customHtml, it has no usable body — sending would deliver
+     * uncompiled <mjml> markup or an empty email. This method blocks that bad
+     * state at save time so it is caught before send. Unpublished/draft emails
+     * are allowed through so the user can save and open the builder later.
+     *
+     * @throws MjmlThemeEmptyCustomHtmlException
+     */
+    public function validateMjmlThemeHasCustomHtml(Email $email): void
+    {
```
Comment (andersonjeccel) [inline app/bundles/EmailBundle/Model/EmailModel.php:299]: This validation is only called from the HTML/API controllers, so it is not enforced at the model save boundary. `EmailModel::saveEntity()` still calls `parent::saveEntity()` without invoking it, and `saveEntities()` persists directly through `emailRepository`; imports, bulk saves, and other callers can therefore still persist a published MJML email with empty `customHtml`. Please enforce the invariant in the model save path (while preserving the draft exception), rather than relying only on controller hooks.

**Review by andersonjeccel — COMMENTED** [review]
This validation is only called from the HTML/API controllers, so it is not enforced at the model save boundary. `EmailModel::saveEntity()` still calls `parent::saveEntity()` without invoking it, and `saveEntities()` persists directly through `emailRepository`; imports, bulk saves, and other callers can therefore still persist a published MJML email with empty `customHtml`. Please enforce the invariant in the model save path (while preserving the draft exception), rather than relying only on controller hooks.

**Comment by santhiprakash** [comment]
Thanks @andersonjeccel — both points addressed in `913ca0dcfd`:

**1. Enforce at the model save boundary**
`validateMjmlThemeHasCustomHtml()` is now called from both `EmailModel::saveEntity()` and `EmailModel::saveEntities()`, so imports, bulk saves, and any other caller that bypasses the controllers can no longer persist a published MJML email with empty `customHtml`. The draft exception (`!$email->getIsPublished()`) is preserved — unpublished emails can still be saved and opened in the builder later. The controller hooks remain for nicer UI/API error messages (flash / 422), but the model is now the enforcement boundary.

The earlier attempt in `c60624fb3e` put validation in `saveEntity()` without the published-only guard, which broke 17 fixture tests that legitimately create draft MJML emails with empty `customHtml`. With the `!$email->getIsPublished()` early return in place, all 782 EmailBundle tests pass.

**2. Narrow the catch clause**
Changed `catch (\Throwable)` to `catch (Twig\Error\Error)` so a genuinely broken theme (syntax/runtime/loader error) is not silently swallowed — only Twig rendering failures skip validation, and `MailHelper::setEmail()` still handles the send-time fallback.

Verification:
```
ddev exec php bin/phpunit -c app/phpunit.xml.dist app/bundles/EmailBundle/Tests
→ OK (782 tests, 4483 assertions)
ddev composer phpstan  → OK
ddev composer rector -- --dry-run  → OK
ddev composer cs  → clean
```

File: app/bundles/EmailBundle/Model/EmailModel.php (hunk 4 above)
Comment (santhiprakash) [reply app/bundles/EmailBundle/Model/EmailModel.php:314]: Addressed in `913ca0dcfd` — narrowed to `catch (Twig\Error\Error)` so only Twig rendering failures skip validation.

File: app/bundles/EmailBundle/Model/EmailModel.php (hunk 5 above)
Comment (santhiprakash) [reply app/bundles/EmailBundle/Model/EmailModel.php:299]: Addressed in `913ca0dcfd` — validation is now enforced in both `saveEntity()` and `saveEntities()` at the model boundary, with the draft exception preserved.

**Comment by peculiaruc** [comment]
LGTM

**Comment by santhiprakash** [comment]
Merged the latest `7.x` into this branch to clear the `BEHIND` merge status.\n\nRe-ran the targeted EmailBundle tests locally after the merge:\n\n- `MailHelperTest` \u2014 3/3 tests pass (18 assertions)\n- `EmailControllerFunctionalTest::testSaveRejectsMjmlThemeWithEmptyCustomHtml` \u2014 passes (4 assertions)\n\nCI is running on the updated branch; will keep an eye on it.

**Review by peculiaruc — APPROVED** [review]

**Comment by santhiprakash** [comment]
Merged the latest `7.x` (`84a6281896`) to clear the behind-base-branch status — no file overlap with this PR's EmailBundle/ProjectBundle changes, clean merge.

Re-ran the targeted tests on the merged head (`287de2df21`):

- `MailHelperTest` (MJML cases) — 2/2 pass, 14 assertions
- `EmailControllerFunctionalTest::testSaveRejectsMjmlThemeWithEmptyCustomHtml` — passes, 4 assertions

Ready for review @andersonjeccel.

**Comment by santhiprakash** [comment]
CI on the rebased head failed for reasons unrelated to this PR; pushed an empty commit (7e9232c4e3) to retrigger.

Triage of run 32035733602:
- `PHPUnit 8.2 mariadb:10.11`: single failure in `AssetApiControllerFunctionalTest::testCreateNewRemoteAssetWithInvalidFile` ("Not allowed php MIME type" data set) — the test fetches `https://raw.githubusercontent.com/mautic/mautic/7.x/index.php` and the runner could not resolve its mimetype. Network flake; AssetBundle is untouched by this PR.
- `e2e-tests`: ddev environment setup failure — a composer hook exited 255 and `vendor/autoload.php` was never installed, so Mautic install and codecept could not start.
- `PHPUnit 8.2 mysql:8.4` / `PHPUnit 8.5 mysql:9.2`: canceled via fail-fast; the 8.5 log shows all 6700 tests passing at cancellation.

Upstream `7.x` CI is green on the exact merged commit (84a628189), and this PR's diff is EmailBundle plus ProjectBundle test fixtures only.

Update: the retriggered run also failed at job setup — GitHub is having a [major partial outage](https://www.githubstatus.com/) since 13:40 UTC (codeload returns 429 when runners download actions, e.g. `shivammathur/setup-php`). The original run's network flakes (13:36-13:44 UTC) coincide with the incident start. Will retrigger once GitHub recovers.

**Comment by santhiprakash** [comment]
@peculiaruc the only thing between this PR and merge now is a re-approval — your Aug 14 approval was invalidated by pushes that were all base maintenance, none of them code changes:

- Aug 15: rebase onto a newer `7.x` tip
- Aug 17: merged `7.x` (`84a6281896`) to clear the behind-base-branch flag — no overlap with this PR's EmailBundle/ProjectBundle files
- 4 empty commits just to retrigger CI after transient runner/API flakes

I verified the final patch is byte-identical to the one you approved (same 11 files, +292/−2), and all 16 CI checks pass on the current head `cbfe0d5314`. Could you re-approve when you have a moment? Apologies for the churn from the CI retries.

**Comment by santhiprakash** [comment]
Quick correction to my last comment, which had the gate analysis wrong: @peculiaruc's Aug 14 approval is still in place — the later rebase and empty CI-retrigger pushes did not dismiss it, so no re-approval is needed there. Apologies for the noise.

The one requirement actually outstanding is the codeowner review: `**/app/bundles/EmailBundle` maps to @mabumusa1 in CODEOWNERS, and that's the approval this PR is still blocked on. @mabumusa1 whenever you have a moment — CI is fully green on the current head (`cbfe0d5314`, all 16 checks) and all earlier review feedback has been addressed.

**Comment by peculiaruc** [comment]
Noted

**Comment by peculiaruc** [comment]
> Quick correction to my last comment, which had the gate analysis wrong: @peculiaruc's Aug 14 approval is still in place — the later rebase and empty CI-retrigger pushes did not dismiss it, so no re-approval is needed there. Apologies for the noise.
> 
> The one requirement actually outstanding is the codeowner review: `**/app/bundles/EmailBundle` maps to @mabumusa1 in CODEOWNERS, and that's the approval this PR is still blocked on. @mabumusa1 whenever you have a moment — CI is fully green on the current head (`cbfe0d5314`, all 16 checks) and all earlier review feedback has been addressed.

Noted

**Comment by santhiprakash** [comment]
Following up on the codeowner gate with two data points that might help unblock this one and #17002:

- @mabumusa1's public review history in mautic shows no activity since early June (last review-bearing PR update June 8, nothing in his public events since).
- The last EmailBundle PR, #16941, merged August 4 on @escopecz's approval — so this gate has been satisfied by a maintainer approval before.

Happy to keep waiting for @mabumusa1 if that's the right call — but if he's away for a while, would an admin be willing to approve/merge this one (and possibly #17002, which touches EmailBundle's ReportSubscriber, is approved by @andersonjeccel, and is also CI-green)?

Current state of this PR: head `ce3d7144f7`, 15/15 checks green, 0 commits behind 7.x.

**Comment by escopecz** [comment]
@santhiprakash anyone from the community can make the second review. It's not physically possible for me to review and test each and every PR. I have to choose my battles. I know this is frustrating to rely on volunteers. Sometimes it helps to lobby for a second reviewer on Slack or ask the issue reporter for help with testing.

**Comment by santhiprakash** [comment]
Thanks for the guidance — I understand. I'll look into getting a second community review via Slack or by asking the issue reporter to help test.

File: app/bundles/EmailBundle/Helper/MailHelper.php (hunk 1 above)
Comment (santhiprakash) [reply app/bundles/EmailBundle/Helper/MailHelper.php:1277]: Done in `61d817b612` — when the MJML fallback is skipped and `plainText` is also empty, `setEmail()` now appends to `\$this->errors`, so it returns `false` and the send aborts instead of delivering an empty body (see the hard-fail block right below the warning).

File: app/bundles/EmailBundle/Tests/Helper/MailHelperTest.php (hunk 2 above)
Comment (santhiprakash) [reply app/bundles/EmailBundle/Tests/Helper/MailHelperTest.php:1394]: Done in `61d817b612` — the test now asserts the logger receives the warning exactly once (`MailHelperTest` ~L1421), and `testSetEmailFailsWhenMjmlThemeSkippedAndNoPlainText` (~L1444) covers empty `customHtml` + empty `plainText`, asserting `setEmail()` returns `false`.

File: app/bundles/EmailBundle/Tests/Controller/EmailControllerFunctionalTest.php (hunk 3 above)
Comment (santhiprakash) [reply app/bundles/EmailBundle/Tests/Controller/EmailControllerFunctionalTest.php:1219]: Done in `61d817b612` — added `EmailControllerFunctionalTest::testSaveRejectsMjmlThemeWithEmptyCustomHtml` (~L1223): saves an MJML-theme email with empty `customHtml` via the API, expects 422 with a clear validation message, and asserts `<mjml>` never reaches the body/MIME.

**Comment by santhiprakash** [comment]
@andersonjeccel you reviewed this early on and your model-boundary concern drove the main structural change since — would you be open to a second look?

Your Aug 10 point — that controller hooks alone still let `EmailModel::saveEntity()` / `saveEntities()` persist a published MJML email with empty `customHtml` — is now enforced at the model save boundary:

- `saveEntity()` calls `validateMjmlThemeHasCustomHtml()` before persisting (`app/bundles/EmailBundle/Model/EmailModel.php`)
- `saveEntities()` runs the same check per entity before the repository writes
- drafts/unpublished emails are exempt so the builder can still open; an unrenderable theme template falls back to the existing `MailHelper::setEmail()` send-time behavior

Current state: head `89d17f38` is green on all 15 checks (PHPUnit matrix, PHPStan, CS Fixer, Rector, SonarCloud quality gate), @peculiaruc approved Aug 14, and per @escopecz the remaining gate is a second community review. Fixes #16988.

**Comment by santhiprakash** [comment]
Merged the latest `7.x` (`6c4d234119`, the migration-metadata fix from #17127 — no overlap with this PR's EmailBundle changes) to clear the behind-base-branch status. No code changes of our own; CI is re-running on the new head `57c81e356`.

**Comment by redbullpeter** [comment]
I came to this from #17479, which reports that plain text only emails are not possible. While investigating that I found it runs through the same fallback this PR changes, so hoping it would fix it, I tested this branch.

Tested on `5927b6eadb` against a fresh install on `cd58803491`, the merge base with `7.x`, so the only delta was this PR.

It does what it sets out to do. On the base, an MJML theme with empty `customHtml` delivers the blank theme's MJML source as the `text/html` part. With the PR applied that MJML is gone from every case I could produce, so #16988 is closed by this.

There is one case that looks unintended, and it is the one an email with a plain text body lands in. The save path and the send path disagree about whether that email is valid.

The send path allows it. When the fallback is skipped, `setEmail()` records an error only if the plain text is empty too ([MailHelper.php:1282](https://github.com/mautic/mautic/blob/5927b6eadb237adbabd617f3a370a1af16278be9/app/bundles/EmailBundle/Helper/MailHelper.php#L1282)). Empty `customHtml` plus a plain text body is therefore treated as a sendable email.

The save path refuses it. `validateMjmlThemeHasCustomHtml()` returns early only when the template is missing, `customHtml` is non-empty, or the email is unpublished ([EmailModel.php:301](https://github.com/mautic/mautic/blob/5927b6eadb237adbabd617f3a370a1af16278be9/app/bundles/EmailBundle/Model/EmailModel.php#L301)). There is no condition on `plainText`, so a published email with a plain text body and no HTML is rejected.

Measured sequence on the PR branch:

1. `POST /api/emails/new` with `template: mautic_code_mode`, `customHtml: ""`, `plainText: "Plain text body from the plain text field."`, `isPublished: 0`. Saves.
2. `POST /s/emails/sendExample/{id}`, the UI's Send Example. The message that arrives is a single `text/plain` part, no multipart, body exactly the plain text.
3. `PATCH /api/emails/{id}/edit` with `{"isPublished": 1}` on that same email. `422`, `The "mautic_code_mode" theme uses MJML which must be compiled into customHtml before saving. Please open the email in the builder to compile the template.`

Step 3 returns the same `422` whether the email carries a plain text body or not, and the same for `template: blank` as for code mode. So the validator does not distinguish the case the send path allows.

That matters more than it might look, because publishing is not optional. Segment sends only pick up published emails ([`EmailRepository::getPublishedBroadcastsIterable()`](https://github.com/mautic/mautic/blob/5927b6eadb237adbabd617f3a370a1af16278be9/app/bundles/EmailBundle/Entity/EmailRepository.php#L844)). So an email with empty `customHtml` and a plain text body can be saved as a draft and test sent, but it can never be published, and therefore can never go to a list.

I measured the `422` on the API. The UI goes through the same `validateMjmlThemeHasCustomHtml()` call in `newAction()`, `editAction()` and `cloneAction()`, which I read rather than drove.

The two new tests do not disagree with each other; the cell where they would is just not covered. [`testSetEmailSkipsMjmlThemeFallbackWhenCustomHtmlEmpty`](https://github.com/mautic/mautic/blob/5927b6eadb237adbabd617f3a370a1af16278be9/app/bundles/EmailBundle/Tests/Helper/MailHelperTest.php#L1393) sets `plainText` and asserts `setEmail()` returns `true`, and [`testSaveRejectsMjmlThemeWithEmptyCustomHtml`](https://github.com/mautic/mautic/blob/5927b6eadb237adbabd617f3a370a1af16278be9/app/bundles/EmailBundle/Tests/Controller/EmailControllerFunctionalTest.php#L1223) posts no `plainText` key at all. Published plus plain text set is never exercised.

Step 2 uses Send Example, which does not append the tracking pixel, so I expected a real send to differ. It does on default settings, where [`setBody()`](https://github.com/mautic/mautic/blob/5927b6eadb237adbabd617f3a370a1af16278be9/app/bundles/EmailBundle/Helper/MailHelper.php#L808) appends the pixel and the message becomes `multipart/alternative` with an HTML part holding just the ``. But `mailer_append_tracking_pixel` is an admin setting in Email Settings, not a code path. With it switched off and this branch applied, a real send to a contact comes out as a single `text/plain` part:

```
Content-Type: text/plain; charset=utf-8
Content-Transfer-Encoding: quoted-printable

Plain text body from the plain text field.
```

No multipart and no HTML part. Switching that setting off on the merge base, without this PR, makes no difference at all, because the fallback still fills the HTML body with the theme's MJML and the message stays `multipart/alternative`.

So this branch plus an existing setting already produces a genuinely plain text email. The only thing in the way is the save check, which stops that email being published.

If the intent is that a plain text only email stays possible, the smallest change I can see is adding `|| !empty($email->getPlainText())` to that early return.

I applied that one line locally to check it actually does the job. The email that returned `422` at step 3 then publishes. With the tracking pixel off, a segment broadcast to it goes out as a single `text/plain` part:

```
Content-Type: text/plain; charset=utf-8
Content-Transfer-Encoding: quoted-printable

Segment plain text body.
```

That is `mautic:broadcasts:send` reporting 1 sent, 0 failed, through the normal segment path rather than a test send. A functional test posting `plainText` with `isPublished: 1` and expecting a save rather than a `422` would cover the gap in the suite.
